x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/threat_enrichment_factory.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { CreateEventSignalOptions, GetThreatListOptions } from './types';
+import type { SignalSourceHit } from '../types';
+import { getThreatList } from './get_threat_list';
+import { enrichSignalThreatMatchesFromSignalsMap } from './enrich_signal_threat_matches';
+import { type SignalsQueryMap } from './get_signals_map_from_threat_index';
+
+interface ThreatEnrichmentFactoryOptions {
+  threatIndicatorPath: CreateEventSignalOptions['threatIndicatorPath'];
+  signalsQueryMap: SignalsQueryMap;
+  threatFilters: CreateEventSignalOptions['threatFilters'];
+  threatSearchParams: Omit<GetThreatListOptions, 'searchAfter'>;
+}
+
+/**
+ * returns threatEnrichment method used events-first search
+ */
+export const threatEnrichmentFactory = ({
+  signalsQueryMap,
+  threatIndicatorPath,
+  threatFilters,
+  threatSearchParams,
+}: ThreatEnrichmentFactoryOptions) => {
+  const threatEnrichment = (signals: SignalSourceHit[]): Promise<SignalSourceHit[]> => {
+    const getThreats = async () => {
+      const threatIds = signals
+        .map((s) => s._id)
+        .reduce<string[]>((acc, id) => {
+          return [
+            ...new Set([
+              ...acc,
+              ...(signalsQueryMap.get(id) ?? []).map((threatQueryMatched) => threatQueryMatched.id),
+            ]),
+          ];
+        }, [])
+        .flat();
+
+      const matchedThreatsFilter = {
+        query: {
+          bool: {
+            filter: {
+              ids: { values: threatIds },
+            },
+          },
+        },
+      };
+
+      const threatResponse = await getThreatList({
```

Comment:
Should we here specify per page here (9k for example)?
Because looks like we query only 1k threats, but there can be more?

Reply:
this part basically replicates current behaviour in `8.6`, where max of 1,000 threats is fetched before building the enrichment.

I don't know whether this was conscious decision to avoid loading large number of documents in memory(with potentially large number of fields depends on `  _source: [`${threatIndicatorPath}.*`, 'threat.feed.*'],` in threat params) or just overlook.

There is some [hard limit of 10,000 items per page](https://github.com/elastic/kibana/blob/8.6/x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/get_threat_list.ts#L18-L21) though.
Do you think it would be safe to load 9,000 items in one request with no control on size of the document?

We do fetch 9,000 threats in one request on initial phase of the execution. But it returns[ only indicator mapping fields](https://github.com/elastic/kibana/blob/8.6/x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/create_threat_signals.ts#L122-L125), which response is significantly smaller

Reply:
I think before it was like:
 `const calculatedPerPage = perPage ?? INDICATOR_PER_PAGE;`
And `perPage` was 9000 [I think](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/signals/executors/threat_match.ts#L72)

So, it worked before for this amount, at least I don't remember any client SDH about that.

Right now we can potentially miss some threat indicators for matches, and some alerts can't be without matches at all.
In this trade-off, I lean more toward keeping 9k as a limit.

Reply:
> I think before it was like:
const calculatedPerPage = perPage ?? INDICATOR_PER_PAGE;
And perPage was 9000 [I think](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/signals/executors/threat_match.ts#L72)

yes, that's where this part is coming from:

> We do fetch 9,000 threats in one request on initial phase of the execution. But it returns[ only indicator mapping fields](https://github.com/elastic/kibana/blob/8.6/x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/create_threat_signals.ts#L122-L125), which response is significantly smaller

It fetches, 9,000 batch of threat documents, but only indicator mapping fields. So, the size of each document is small.

But, when it fetches threats to make enrichments, it uses [default page size as 1,000 ](https://github.com/elastic/kibana/blob/8.6/x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/build_threat_enrichment.ts#L40) as `perPage` is `undefined` in that call. And it returns, the whole section of document, defined in query parameters
```
  _source: [${threatIndicatorPath}.*, 'threat.feed.*'],
```
for that particular call, which fetches threats for enrichments
So, this implementation is essentially kept in the same way as in 8.6.

While, I agree, it can miss a lot of potential threats my concern is around fetching of large number of documents.
Which could result:
- large memory consumption
- too large content of response that causes rule execution error. [Corresponding issue](https://github.com/elastic/kibana/issues/150041)

> So, it worked before for this amount, at least I don't remember any client SDH about that.
It worked for 1,000 for threats first search

Although it seems, 9,000 threats should not cause the issue as in [150041](https://github.com/elastic/kibana/issues/150041)

@marshallmain, what are you thoughts on this?